File: SortingAlgorithms/SortingAlgorithms/BucketSort.cpp
#include "BucketSort.hpp"
#include <vector>


static int getBucketIndex(float val, int bucketCap) {
	return val * bucketCap;
}

static void bubbleSortModified(std::vector<float>& arr, int size) {
	bool swapped;
	do {
		swapped = false;
		for (int i = 1; i <= size; i++) {
			if (arr[i - 1] > arr[i]) {
				std::swap(arr[i - 1], arr[i]);
				swapped = true;
			}
		}
	} while (swapped);
}

void bucketSort(float arr[], int size, int bucketsNum , int bucketCap) {
	bucketsNum = size;

	std::vector<std::vector<float>> buckets(bucketsNum);

	for (int i = 0; i < size; i++) {
		int index = getBucketIndex(arr[i], bucketCap);
		buckets[index].push_back(arr[i]);
	}

	for (int i = 0; i < bucketsNum; i++) {
		if (buckets[i].size() != 0) bubbleSortModified(buckets[i], buckets[i].size() - 1);
	}

	int j = 0;

	for (int i = 0; i < bucketsNum; i++) {
		if (buckets[i].size() != 0) {
			for (auto x : buckets[i]) {
				arr[j] = x;
				j++;
			}
		}
	}
}

File: SortingAlgorithmsPython/bucketSort.py
import bubbleSort

def get_bucket_index(val, bucket_capacity):
    return int(val * bucket_capacity)

def bucket_sort(arr, buckets_num, bucket_capacity):
    buckets = [[] for _ in range(buckets_num)]

    for i in range(len(arr)):
        index = get_bucket_index(arr[i], bucket_capacity)
        buckets[index].append(arr[i])

    for i in range(buckets_num):
        if buckets[i] != None: bubbleSort.bubble_sort(buckets[i])

    j = 0
    for i in range(buckets_num):
        if buckets[i] != None:
            for x in buckets[i]:
                arr[j] = x
                j += 1